### PR TITLE
[feature](nereids) support like and regexp predicate

### PR DESCRIPTION
**Pull request by yinzhijian**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

support like and regexp predicate for nereids.
for example:
select * from t1 where k1 like 'xxx' and k2 regexp '^sa'

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -576,6 +582,38 @@ public Expression withBetween(PredicateContext ctx, Expression e) {
...
+     * @param e Expression
+     * @return Expression
+     */
+    public Expression withLike(PredicateContext ctx, Expression e) {
+        boolean isNotLike = ctx.NOT() != null ? true : false;
+        LikePredicate likePredicate = new LikePredicate(
+                e,
+                expression(ctx.pattern)
+        );
+        return isNotLike ? new Not(likePredicate) : likePredicate;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:598]: in another pr, we refactor this and generate process in withPredicate

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 2]
```diff
@@ -178,6 +178,7 @@ booleanExpression
 
 predicate
     : NOT? kind=BETWEEN lower=valueExpression AND upper=valueExpression
+    | NOT? kind=(LIKE | REGEXP) pattern=valueExpression
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:181]: we have RLIKE in DorisLexer.g4 represent both 'RLIKE' and 'REGEXP'

Comment (yinzhijian) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:181]: DORIS original optimizer does not support RLIKE, the new optimizer should be consistent with it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 1 above)
Comment (yinzhijian) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:598]: which pr？

Comment (yinzhijian) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Like.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public class Like<LEFT_CHILD_TYPE extends Expression, RIGHT_CHILD_TYPE extends Expression>
+        extends StringRegexPredicate<LEFT_CHILD_TYPE, RIGHT_CHILD_TYPE> {
+
+    public Like(LEFT_CHILD_TYPE left, RIGHT_CHILD_TYPE right) {
+        super(NodeType.LIKE, left, right);
+    }
+
+    @Override
+    public boolean nullable() throws UnboundException {
+        return left().nullable() || right().nullable();
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Like.java:39]: Maybe only ```left().nullable()``` ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Regexp.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+public class Regexp<LEFT_CHILD_TYPE extends Expression, RIGHT_CHILD_TYPE extends Expression>
+        extends StringRegexPredicate<LEFT_CHILD_TYPE, RIGHT_CHILD_TYPE> {
+
+    public Regexp(LEFT_CHILD_TYPE left, RIGHT_CHILD_TYPE right) {
+        super(NodeType.REGEXP, left, right);
+    }
+
+    @Override
+    public boolean nullable() throws UnboundException {
+        return left().nullable() || right().nullable();
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Regexp.java:39]: same as like

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**